src/main/scala/fi/vm/sade/hakurekisteri/integration/cache/RedisUpdateConcurrencyHandler.scala

```diff
@@ -0,0 +1,129 @@
+        }
+        newClientPromise.future
+      }.recoverWith { case e =>
+        logger.error(s"Error when obtaining lock $lockKey", e)
+        Future.failed(e)
+      }
+    }
+  }
+
+  private def storePromiseForThisRequest(key: K): (Promise[Option[T]], Boolean) = {
+    logger.debug("Adding new client promise:")
+    val updateFunction = new BiFunction[K, List[Promise[Option[T]]], List[Promise[Option[T]]]] {
+      override def apply(key: K, promises: List[Promise[Option[T]]]): List[Promise[Option[T]]] = {
+        val myPromise = Promise[Option[T]]
+        if (promises == null) {
+          List(myPromise)
+        } else {
+          myPromise :: promises
+        }
+      }
+    }
+    val (newClientPromise :: otherPromises): Seq[Promise[Option[T]]] = waitingPromises.compute(key, updateFunction)
+
+    val numberWaiting = otherPromises.size + 1
+    logger.debug(s"Waiting after adding: $numberWaiting")
+    if (numberWaiting > limitOfWaitingClientsToLog) {
+      logger.warn(s"Already $numberWaiting clients waiting for value of $key")
+    }
+    (newClientPromise, otherPromises.nonEmpty)
+  }
+
+  private def retrieveNewvalueWithLoader(key: K,
+                                         loader: K => Future[Option[T]],
+                                         loadedValueStorer: (K, Future[T]) => Future[_],
+                                         lockKey: String): Unit = {
+    val loadingFuture = loader(key)
+    loadingFuture.onComplete { result =>
+      if (result.isSuccess) {
+        result.get match {
+          case Some(foundItem) =>
+            loadedValueStorer(key, Future.successful(foundItem)).onComplete { _ =>
+              resolvePromisesWaitingForValueFromCache(key, result, failIfNobodyWaiting = false)
+              removeCacheUpdateLock(lockKey)
+            }
+          case None =>
+            removeCacheUpdateLock(lockKey)
+        }
+      } else {
+        removeCacheUpdateLock(lockKey)
+      }
+      resolvePromisesWaitingForValueFromCache(key, result)
+    }
+  }
+
+  private def resolvePromisesWaitingForValueFromCache(key: K, result: Try[Option[T]], failIfNobodyWaiting: Boolean = true): Unit = {
+    val updateFunction = new BiFunction[K, List[Promise[Option[T]]], List[Promise[Option[T]]]] {
+      override def apply(key: K, promisesWaitingForResult: List[Promise[Option[T]]]): List[Promise[Option[T]]] = {
+        if (promisesWaitingForResult != null) {
+          logger.debug(s"Resolving promises for key $key: ${promisesWaitingForResult.size}")
+          promisesWaitingForResult.foreach { _.complete(result) }
+        } else {
+          if (failIfNobodyWaiting) {
+            throw new IllegalStateException(s"Nobody waiting for results of $key , got result $result")
+          } else {
+            logger.debug(s"Nobody waiting for results of $key, got result $result")
+          }
+        }
+        null
+      }
+    }
+
+    waitingPromises.compute(key, updateFunction)
```

Comment:
Tässä itse asiassa pitäis toimia se edellinen `.remove`:een perustunut ratkasu about sellassenaan.